### DISCUSSION TITLE
zero copy between ByteBuf(netty) and protobuf

### DISCUSSION THREAD

**Discussion by LinHuiG** [netty/netty#13800] [Q&A]
hello,we used protobuf in netty,it powerful but we need read bytes from ByteBuf before decode or encode to bytes and then write to ByteBuf ,this once copy between ByteBuf(netty) and protobuf  make me puzzle.
i found method writeTo and readFrom in protoBuf,but i dont konw how to use it in netty to achieve zero copy,anyone can tell me ?
very thanks!

**Comment by normanmaurer** [comment]
I think there are methods that take a `ByteBuffer` in protobuf. You could use these as you can get a `ByteBuffer` via `nioBuffer()`

**Reply by LinHuiG** [reply]
i tried it but not work, one `ByteBuf `could has multiple `ByteBuffer`,we may need `nioBuffers()` if we want to get `ByteBuffer ` from `ByteBuf`
